api/handlers.go

```diff
@@ -0,0 +1,156 @@
+	router.HandleFunc("POST /certificate_requests", PostCertificateRequest(env))
+	router.HandleFunc("GET /certificate_requests/{id}", GetCertificateRequest(env))
+	router.HandleFunc("DELETE /certificate_requests/{id}", DeleteCertificateRequest(env))
+	router.HandleFunc("POST /certificate_requests/{id}/certificate", PostCertificate(env))
+
+	v1 := http.NewServeMux()
+	v1.HandleFunc("GET /status", HealthCheck)
+	v1.Handle("/api/v1/", http.StripPrefix("/api/v1", router))
+
+	return logging(v1)
+}
+
+// the health check endpoint simply returns a 200
+func HealthCheck(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(200) //nolint:errcheck
+}
+
+// GetCertificateRequests returns all of the Certificate Requests
+func GetCertificateRequests(env *Environment) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		certs, err := env.DB.RetrieveAll()
+		if err != nil {
+			logError(err.Error(), 500, w)
+			return
+		}
+		body, err := json.Marshal(certs)
+		if err != nil {
+			logError(err.Error(), 500, w)
+			return
+		}
+		if _, err := w.Write(body); err != nil {
+			logError(err.Error(), 500, w)
+		}
+	}
+}
+
+// PostCertificateRequest creates a new Certificate Request, and returns the id of the created row
+func PostCertificateRequest(env *Environment) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		csr, err := io.ReadAll(r.Body)
+		if err != nil {
+			logError(err.Error(), 500, w)
+			return
+		}
+		id, err := env.DB.Create(string(csr))
+		if err != nil {
+			if strings.Contains(err.Error(), "UNIQUE constraint failed") {
+				logError("given csr already recorded", 400, w)
+				return
+			} else {
+				logError(err.Error(), 400, w)
+				return
+			}
+		}
+		w.WriteHeader(201)
+		if _, err := w.Write([]byte(strconv.FormatInt(id, 10))); err != nil {
+			logError(err.Error(), 500, w)
+		}
+	}
+}
+
+// GetCertificateRequests receives an id as a path parameter, and
+// returns the corresponding Certificate Request
+func GetCertificateRequest(env *Environment) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		id := r.PathValue("id")
+		cert, err := env.DB.Retrieve(id)
+		if err != nil {
+			if err.Error() == "sql: no rows in result set" {
+				logError(err.Error(), 400, w)
+				return
+			}
+			logError(err.Error(), 500, w)
+			return
+		}
+		body, err := json.Marshal(cert)
+		if err != nil {
+			logError(err.Error(), 500, w)
+			return
+		}
+		if _, err := w.Write(body); err != nil {
+			logError(err.Error(), 500, w)
+		}
+	}
+}
+
+// DeleteCertificateRequest handler receives an id as a path parameter,
+// deletes the corresponding Certificate Request, and returns a 204 on success
+func DeleteCertificateRequest(env *Environment) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		id := r.PathValue("id")
+		err := env.DB.Delete(id)
+		if err != nil {
+			logError(err.Error(), 500, w)
```

Comment:
Do you know if an error is returned by the DB if we try to delete something that does not exist? In that case, we should not return a 500, but something either a success, or something in the 4xx range.

Reply:
No, if nothing is deleted there is no error. It's still a good idea to differentiate this, so I've added a custom error.